src/sdk/namespace_fs.js

```diff
@@ -199,6 +240,62 @@ const dir_cache = new LRUCache({
     max_usage: config.NSFS_DIR_CACHE_MAX_TOTAL_SIZE,
 });
 
+/**
+ * @typedef {{
+ *  time: number,
+ *  stat: nb.NativeFSStats,
+ *  usage: number,
+ *  sorted_entries?: fs.Dirent[],
+ * }} ReaddirVersionsCacheItem
+ * @type {LRUCache<object, string, ReaddirCacheItem>}
+ */
+const versions_dir_cache = new LRUCache({
+    name: 'nsfs-versions-dir-cache',
+    make_key: ({ dir_path }) => dir_path,
+    load: async ({ dir_path, fs_context }) => {
+        const time = Date.now();
+        const stat = await nb_native().fs.stat(fs_context, dir_path);
+        const version_path = dir_path + "/" + HIDDEN_VERSIONS_PATH;
+        let ver_dir_stat_size;
+        let is_version_path_exists = false;
+        try {
+            const ver_dir_stat = await nb_native().fs.stat(fs_context, version_path);
+            ver_dir_stat_size = ver_dir_stat.size;
+            is_version_path_exists = true;
+        } catch (err) {
+            if (err.code === 'ENOENT') {
+                dbg.log0('NamespaceFS: Version dir not found, ', version_path);
+            } else {
+                throw err;
+            }
+            ver_dir_stat_size = 0;
+        }
+        let sorted_entries;
+        let usage = config.NSFS_DIR_CACHE_MIN_DIR_SIZE;
+        if (stat.size + ver_dir_stat_size <= config.NSFS_DIR_CACHE_MAX_DIR_SIZE) {
+            const latest_versions = await nb_native().fs.readdir(fs_context, dir_path);
+            if (is_version_path_exists) {
+                const old_versions = await nb_native().fs.readdir(fs_context, version_path);
+                const entries = latest_versions.concat(old_versions);
+                sorted_entries = entries.sort(sort_entries_by_name_and_time);
+            } else {
+                sorted_entries = latest_versions.sort(sort_entries_by_name);
+            }
+            /*eslint no-unused-expressions: ["error", { "allowTernary": true }]*/
+            for (const ent of sorted_entries) {
+                usage += ent.name.length + 4;
+            }
+        }
+        return { time, stat, sorted_entries, usage };
+    },
+    validate: async ({ stat }, { dir_path, fs_context }) => {
+        const new_stat = await nb_native().fs.stat(fs_context, dir_path);
+        return (new_stat.ino === stat.ino && new_stat.mtimeNsBigint === stat.mtimeNsBigint);
```

Comment:
do we need here validation of ver_dir_stat_size as well?

Reply:
May be we can consider it in separate PR to optimize it ? I will open up PR for this.

Reply:
I'm fine with handling it in a different PR, as long as there are no regressions